### PR TITLE
Feature invalid version recognition

### LINKED ISSUE #2218: 1.3 invalid version recognition
HNQIS app must recognize an invalid version in 2 situations:
- On login, if trying to perform a login to a DHIS2 > 2.25 then it should show an alert saying that version is not supported
- On push, HNQIS must check before every push that server version didn't change and is still 2.25. In case it's not, then an alert is shown recommending to upgrade the app from GPlay

### PR DESCRIPTION
**Pull request by idelcano**
### :pushpin: References
* **Issue:** close #2218

###   :gear: branches 
**app**: 
       Origin: feature-invalid_version_recognition Target: v1.3_hnqis
**bugshaker-android**: 
       Origin: downgrade_gradle_version 
**dhis2-light-sdk**:
       Origin: development
**EyeSeeTea-SDK**: 
       Origin: development
**SDK**: 
       Origin: 2.25_EyeSeeTea

### :tophat: What is the goal?

HNQIS app must recognize an invalid version in 2 situations:

On login, if trying to perform a login to a DHIS2 > 2.25 then it should show an alert saying that version is not supported
On push, HNQIS must check before every push that server version didn't change and is still 2.25. In case it's not, then an alert is shown recommending to upgrade the app from GPlay and cancel alarm.

### :memo: How is it being implemented?

- I have copied from pictureapp the mockwebserver and file reader test utils.
- I have updated the gradle versions to avoid compilation errors. (powerMockVersion is required by dhis sdk).
- I have removed old expresso tests
- I have added pushusecase and loginusecase tests to test the server version errors with mocked server responses read from json files.
- I have created a "visible" variable in dashboardactivity to know if the activity is visible before shown the server version error dialog.
- I have updated the push and login use case to use executors, and I added a server version validation, and server version error callbacks and treatments.
- I have added an API request to know the server version using ServerInfoDataSource.
- I have refactored the API PullDhisApiDataSource  class to make equal to v1.4.

### :boom: How can it be tested?
You can test the invalid server responses changing the last valid server version on resource "api_minimal_server_version" on donottranslate.xml

 **Use case 1:** -  Realize login with url https://data-psi-mis.org and max_compatible_server_version = 24. You should see a dialog with server version error.
 **Use case 2:** - Realize login with url https://data-psi-mis.org and max_compatible_server_version = 25. You should login ok
 **Use case 3:** - Realize login with url https://data-psi-mis.org and max_compatible_server_version = 25 and push shoud not show any message to user.
 **Use case 4:** - After login with max_compatible_server_version to 25, change the max_compatible_server_version to 24, and run the app. You should see a dialog recommending update the app. And the push will stop. If you close and open the app, a new push will show the same dialog.
 **Use case 5:** - Run instrumental tests
 **Use case 6:** - Run unitary test

### :floppy_disk: Requires DB migration?

- [ ] Nope, we can just merge this branch.
- [ ] Yes, but we need to apply it before merging this branch.
- [x] Yes, it's already applied.

### :art: UI changes?

- [x] Nope, the UI remains as beautiful as it was before!
- [ ] Yeap, here you have some screenshots-

### CODE REVIEW

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCaseShould.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+            @Override
+            public void onServerVersionError() {
+                fail("onServerVersionError");
+            }
+        });
+    }
+
+    @Test
+    public void return_on_server_version_error_when_server_version_is_lower_than_last_valid_server() throws Exception {
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCaseShould.java]: The name of this test is right but the test is wrong.
The test code should be according to test name.
minimalVersion = 26 and server version = 25 and the test should fail

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCaseShould.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                fail("onNetworkError");
+            }
+
+            @Override
+            public void onServerVersionError() {
+                Assert.assertTrue(true);
+            }
+        });
+    }
+
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCaseShould.java:167]: Please create a new test with name:
return__on_login_success_when_server_version_is_greather_than_last_valid_server()
The code should be minimalVersion = 25 and server version = 26 and the use case should work

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+            }
+
+            @Override
+            public void onPushInProgressError() {
+                fail("onLoginSuccess");
+            }
+
+            @Override
+            public void onSurveysNotFoundError() {
+                Assert.assertTrue(true);
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java]: This assert is never invoked

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+    private static final String SYSTEM_INFO_VERSION_25 = "system_info_25.json";
+    private static final String SYSTEM_INFO_VERSION_26 = "system_info_26.json";
+    private static final String AUTH = "auth.json";
+
+    @Rule
+    public MockWebServerRule mockWebServerRule = new MockWebServerRule();
+
+    @Test
+    public void return_on_surveys_not_found_callback_when_server_version_with_invalid_server_version_when_the_user_is_demo_user() throws Exception {
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java]: Review this name

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+            @Override
+            public void onServerVersionError() {
+                fail("onLoginSuccess");
+            }
+        });
+    }
+
+    @Test
+    public void return_on_surveys_not_found_callback_when_server_version_is_equals_than_last_valid_server() throws Exception {
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java]: review this name

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+            @Override
+            public void onServerVersionError() {
+                fail("onLoginSuccess");
+            }
+        });
+    }
+
+    @Test
+    public void return_on_server_version_error_callback_when_server_version_is_high_than_last_valid_server() throws Exception {
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java]: review this name

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            }
+
+            @Override
+            public void onPushInProgressError() {
+                fail("onLoginSuccess");
+            }
+
+            @Override
+            public void onSurveysNotFoundError() {
+                Assert.assertTrue(true);
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java]: This assert is never invoked

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+            @Override
+            public void onServerVersionError() {
+                fail("onLoginSuccess");
+            }
+        });
+    }
+
+    @Test
+    public void return_on_server_version_error_callback_when_server_version_is_high_than_last_valid_server() throws Exception {
+        Credentials credentials = new Credentials(mockWebServerRule.getMockServer().getBaseEndpoint(), "user", "password");
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java:152]: this test should not fail:
If last valid version is 25 and server version is 26 then It's OK

File: app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+                fail("onNetworkError");
+            }
+
+            @Override
+            public void onServerVersionError() {
+                Assert.assertTrue(true);
+            }
+        });
+    }
+
```
Comment (xurxodev) [inline app/src/androidTest/java/org/eyeseetea/malariacare/domain/usecase/PushUseCaseShould.java:201]: Please add a test where the last valid version is 26 and server version is 25 then use case should fail

File: app/src/main/res/values/donottranslate.xml [hunk 10]
```diff
@@ -18,6 +18,7 @@
   ~  along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
   -->
 <resources>
+    <string name="api_minimal_server_version" translatable="false">25</string>
```
Comment (xurxodev) [inline app/src/main/res/values/donottranslate.xml]: minimal server version should be 26

File: app/src/main/java/org/eyeseetea/malariacare/data/IServerInfoDataSource.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package org.eyeseetea.malariacare.data;
+
+import org.eyeseetea.malariacare.domain.entity.ServerInfo;
+
+public interface IServerInfoDataSource {
```
Comment (xurxodev) [inline app/src/main/java/org/eyeseetea/malariacare/data/IServerInfoDataSource.java:5]: This should be a Repository interface in domain

**Review by xurxodev — CHANGES_REQUESTED** [review]
@idelcano I have found many changes to review

**Review by xurxodev — CHANGES_REQUESTED** [review]
@idelcano Running test the behaviour is ok but running app I have found a strange behavior in push validation.
I have change use cases to test in PR:
The use case 3 - after login against server 25 and with max api version 25 too, an upgrade dialog is shown - This is wrong
The use case 4 - after login against server 25 and with max api version 25 too , close app and change max api version to 24 and restart app, the dialog on push is shown twice

**Comment by idelcano** [comment]
@xurxodev  its solved.
Sorry, I committed by error a duplicate call.

**Review by xurxodev — APPROVED** [review]
thanks @idelcano

**Pull request merged**